Support custom 'order by' expression for system log tables

helifu opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
System log tables can now have custom sorting keys.

helifu committed:
Support custom 'order by' expression for system log tables

rschu1ze reviewed (commented):
Please update the [the docs](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings) for the new setting. Affected are (as far as I see, but please double-check) all system tables created by `SystemLogs::SystemLogs`.

helifu commented on src/Interpreters/SystemLog.cpp:
```diff
@@ -152,19 +152,33 @@ std::shared_ptr<TSystemLog> createSystemLog(
+
+        /// ORDER BY expr is necessary.
+        String order_by = config.getString(config_prefix + ".order_by", TSystemLog::getDefaultOrderBy());
+        engine += " ORDER BY (" + order_by + ")";
```
Emm... the pair of parentheses in line174 is a protection to users'  nonstandard configuration. And it fails when  configured as below:
`<order_by>event_time,query_id</order_by>`
If you don't mind, I will remove the parentheses in function `getDefaultOrderBy(),` what do you think?

rschu1ze replied:
Makes sense, okay.

helifu committed:
Remove the duplicate pair of parentheses

helifu commented:
>

rschu1ze commented on src/Interpreters/SystemLog.cpp:
```diff
@@ -152,19 +152,33 @@ std::shared_ptr<TSystemLog> createSystemLog(
+        /// PARTITION expr is not necessary.
+        String partition_by = config.getString(config_prefix + ".partition_by", "toYYYYMM(event_date)");
         if (!partition_by.empty())
+        {
```
Code style: This file is not using { } for single-line if/for/while statements. Could we please keep a consistent code style and omit { }?

helifu replied:
done.

helifu committed:
Add the 'order_by' description for system log tables

rschu1ze commented on programs/server/config.xml:
```diff
@@ -1133,6 +1133,16 @@
+                event_date, type, query_id
+                event_date, event_time, initial_query_id
+
+        <order_by>event_date, event_time, initial_query_id</order_by>
```
What about configuration for `<trace_log>`, `<query_thread_log>` below? Please update them accordingly (if needed).

helifu replied:
No, it's not ncessary, because it's a little bit stupid to add this configuration to every system log table. The template in the first system log table is sufficient :)

rschu1ze reviewed (commented):
If you have time, a test (or an extension of an existing test) would be nice.

Since server parameters are changed, it will likely be an [integration test](https://clickhouse.com/docs/en/development/tests#integration-tests).

helifu commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -1319,12 +1319,13 @@ Queries are logged in the [system.part_log](../../operations/system-tables/part_
+- `database` - Name of the database.
+- `table` - Name of the system table.
+- `partition_by` - [Custom partitioning key](../../engines/table-engines/mergetree-family/custom-partitioning-key.md) for a system table. Can't be used if `engine` defined.
+- `order_by` - [Custom sorting key](../../engines/table-engines/mergetree-family/mergetree.md#order_by) for a system table. Can't be used if `engine` defined.
```
done.

helifu committed:
Update according to review comments

helifu committed:
Add test case for 'order_by' configuration

helifu commented:
> 

test case has been added.

helifu commented on src/Interpreters/SystemLog.cpp:
```diff
@@ -143,25 +143,40 @@ std::shared_ptr<TSystemLog> createSystemLog(
+        String order_by = config.getString(config_prefix + ".order_by", TSystemLog::getDefaultOrderBy());
+        engine += " ORDER BY (" + order_by + ")";
+
+        /// STORAGE POLICY expr is not necessary.
```
done

azat replied:
Thank you! Can you also add a test for this?

helifu replied:
Sure~ :)

helifu replied:
@azat @rschu1ze Could you please share how to run integration tests locally? :)

azat replied:
There is wrapper for this, that can be used like this (this is an example from README):

```sh
/src/ch/clickhouse/tests/integration/runner --binary $HOME/ClickHouse/programs/clickhouse  --odbc-bridge-binary $HOME/ClickHouse/programs/clickhouse-odbc-bridge --base-configs-dir $HOME/ClickHouse/programs/server/ 'test_ssl_cert_authentication -ss'
```

helifu committed:
Add 'settings' expression for system log tables

helifu committed:
Add test case for 'engine','settings' configuration

rschu1ze commented on tests/integration/test_system_logs/test_system_logs.py:
```diff
@@ -0,0 +1,73 @@
+
+node1 = cluster.add_instance(
+    "node1",
+    main_configs=["configs/config_order_by.xml"],
```
Maybe I miss something but I don't find this file in this PR. On the other hand, a new file "system_logs_order_by.xml" was added but it was not referenced anywhere. Did you perhaps mix both files up?

helifu replied:
Ah, wrong indeed :(

robot-clickhouse commented:
<!-- automatic status comment for PR #49778 from helifu/ClickHouse:master2 -->
This is an automated comment for commit 84af8fd5363a1b3a413ee844b2d51c1439565d60 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49778/84af8fd5363a1b3a413ee844b2d51c1439565d60/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

helifu committed:
fix the config file name

helifu committed:
fix a style checking issue

rschu1ze commented:
test_system_logs unfortunately fails (I didn't check closer), @helifu would you like to check or merge from master again?

helifu committed:
Fix failed test cases

test_system_logs/test_system_logs.py::test_system_logs_order_by_expr PASSED   [ 33%]
test_system_logs/test_system_logs.py::test_system_logs_engine_expr PASSED     [ 66%]
test_system_logs/test_system_logs.py::test_system_logs_settings_expr PASSED   [100%]

helifu committed:
Fix a style checking issue

rschu1ze commented:
01676_clickhouse_client_autocomplete and test_system_logs look suspicious.
Fuzzer issue: #50245

Let me restart CI.

helifu committed:
Node2 cannot be started due to a conflict with config.xml merge

helifu commented:
> 01676_clickhouse_client_autocomplete and test_system_logs look suspicious. Fuzzer issue: #50245
> 
> Let me restart CI.

I found the root issue:

- ./integration/test_system_logs/_instances_system_logs/node2/database/preprocessed_configs/config.xml

```
    <query_log>
        <!-- What table to insert data. If table is not exist, it will be created.
             When query log structure is changed after system update,
              then old table will be renamed and new table will be created automatically.
        -->
        <database>system</database>
        <table>query_log</table>
        <!--
            PARTITION BY expr: https://clickhouse.com/docs/en/table_engines/mergetree-family/custom_partitioning_key/
            Example:
                event_date
                toMonday(event_date)
                toYYYYMM(event_date)
                toStartOfHour(event_time)
        -->
        <partition_by>toYYYYMM(event_date)</partition_by>
        <!--
            Table TTL specification: https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree/#mergetree-table-ttl
            Example:
                event_date + INTERVAL 1 WEEK
                event_date + INTERVAL 7 DAY DELETE
                event_date + INTERVAL 2 WEEK TO DISK 'bbb'

        <ttl>event_date + INTERVAL 30 DAY DELETE</ttl>
        -->

        <!--
            ORDER BY expr: https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#order_by
            Example:
                event_date, event_time
                event_date, type, query_id
                event_date, event_time, initial_query_id

        <order_by>event_date, event_time, initial_query_id</order_by>
        -->

        <!-- Instead of partition_by, you can provide full engine expression (starting with ENGINE = ) with parameters,
             Example: <engine>ENGINE = MergeTree PARTITION BY toYYYYMM(event_date) ORDER BY (event_date, event_time) SETTINGS index_granularity = 1024</engine>
          -->

        <!-- Interval of flushing data. -->
        <flush_interval_milliseconds>7500</flush_interval_milliseconds>

        <!-- example of using a different storage policy for a system table -->
        <!-- storage_policy>local_ssd</storage_policy -->
    
        <engine>Engine = MergeTree PARTITION BY event_date ORDER BY event_time TTL event_date + INTERVAL 30 day SETTINGS storage_policy='policy2', ttl_only_drop_parts=1</engine>
    </query_log>
```

- integration/test_system_logs/_instances_system_logs/node2/logs/clickhouse-server.log

```
2023.05.25 15:12:02.902876 [ 8 ] {} <Debug> SystemLog: Creating system.query_log from query_log
2023.05.25 15:12:02.903357 [ 8 ] {} <Error> Application: Caught exception while loading metadata: Code: 36. DB::Exception: If 'engine' is specified for system table, PARTITION BY parameters should be specified directly inside 'engine' and 'partition_by' setting doesn't make sense. (BAD_ARGUMENTS), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000e3c7415 in /usr/bin/clickhouse
1. ? @ 0x00000000139deb58 in /usr/bin/clickhouse
2. ? @ 0x0000000013dcae05 in /usr/bin/clickhouse
3. DB::SystemLogs::SystemLogs(std::shared_ptr<DB::Context const>, Poco::Util::AbstractConfiguration const&) @ 0x0000000013dc4f70 in /usr/bin/clickhouse
4. DB::Context::initializeSystemLogs() @ 0x0000000013097595 in /usr/bin/clickhouse
5. DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0x000000000e45df8d in /usr/bin/clickhouse
6. Poco::Util::Application::run() @ 0x0000000017c67e06 in /usr/bin/clickhouse
7. DB::Server::run() @ 0x000000000e44e31d in /usr/bin/clickhouse
8. Poco::Util::ServerApplication::run(int, char**) @ 0x0000000017c7b539 in /usr/bin/clickhouse
9. mainEntryClickHouseServer(int, char**) @ 0x000000000e44b2e6 in /usr/bin/clickhouse
10. main @ 0x00000000088851cf in /usr/bin/clickhouse
11. __libc_start_main @ 0x00007f83b2542083 in ?
12. _start @ 0x0000000007fb5dae in /usr/bin/clickhouse
 (version 23.5.1.2106)
2023.05.25 15:12:02.903417 [ 8 ] {} <Information> Application: Shutting down storages.
```

helifu committed:
fix a style checking issue

helifu committed:
Add input parameter 'base_config_dir'

helifu commented:
@rschu1ze Seems the failed test cases have nothing to do with the modification in this PR.

rschu1ze commented:
AST Fuzzer: #49998
Integration tests (asan) [4/6]: #50309
ClickHouse Stateless Tests (release, Analyzer): broken in master, unrelated
